compiler/src/iree/compiler/Dialect/Util/Transforms/OptimizeIntArithmetic.cpp

```diff
@@ -186,6 +189,132 @@ struct ConvertUnsignedI64IndexCastProducerToIndex
+    }
+    ArrayRef<Attribute> assumptions = op.getAssumptions().getValue();
+    // Ugly hack to avoid recreating the assumptions array.
+    ArrayRef<ArrayAttr> castAssumptions = ArrayRef(
+        static_cast<const ArrayAttr *>(assumptions.data()), assumptions.size());
+    auto newOp = rewriter.create<Util::AssumeIntOp>(op.getLoc(), newArgs,
+                                                    castAssumptions);
+    SmallVector<Value> replacements(newOp.getResults());
+    for (auto [newRes, oldRes] :
+         llvm::zip_equal(replacements, op.getResults())) {
+      if (newRes.getType() != oldRes.getType()) {
+        newRes = rewriter.create<arith::IndexCastUIOp>(
+            op.getLoc(), oldRes.getType(), newRes);
+      }
+      // Preserve assumption state.
+      auto *oldState = solver.lookupState<IntegerValueRangeLattice>(oldRes);
+      if (oldState) {
+        (void)solver.getOrCreateState<IntegerValueRangeLattice>(newRes)->join(
+            *oldState);
+      }
+    }
+    rewriter.replaceOp(op, replacements);
+    return success();
+  }
+
+  DataFlowSolver &solver;
+};
+
+//===----------------------------------------------------------------------===//
+// scf.for induction variable range narrowing
+// If the induction variable of an scf.for can be represented as an I32,
+// make that change to save on registers etc.
+//===----------------------------------------------------------------------===//
+struct NarrowSCFForIvToI32 : public OpRewritePattern<scf::ForOp> {
+  NarrowSCFForIvToI32(MLIRContext *context, DataFlowSolver &solver)
+      : OpRewritePattern(context), solver(solver) {}
+
+  LogicalResult matchAndRewrite(scf::ForOp forOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = forOp.getLoc();
+    Value iv = forOp.getInductionVar();
+    Type srcType = iv.getType();
+    if (!srcType.isIndex() && !srcType.isInteger(64))
+      return rewriter.notifyMatchFailure(forOp, "IV isn't an index or i64");
+    if (!staticallyLegalToConvertToUnsigned(solver, iv))
+      return rewriter.notifyMatchFailure(forOp, "IV isn't non-negative");
+    if (!staticallyLegalToConvertToUnsigned(solver, forOp.getStep()))
+      return rewriter.notifyMatchFailure(forOp, "Step isn't non-negative");
+    auto *ivState = solver.lookupState<IntegerValueRangeLattice>(iv);
+    if (ivState->getValue().getValue().smax().getActiveBits() > 31)
+      return rewriter.notifyMatchFailure(forOp, "IV won't fit in signed int32");
+
+    Type i32 = rewriter.getI32Type();
+    auto doCastDown = [&](Value v) -> Value {
+      if (srcType.isIndex())
+        return rewriter.create<arith::IndexCastUIOp>(loc, i32, v);
+      else
+        return rewriter.create<arith::TruncIOp>(loc, i32, v);
+    };
+    Value newLb = doCastDown(forOp.getLowerBound());
+    Value newUb = doCastDown(forOp.getUpperBound());
+    Value newStep = doCastDown(forOp.getStep());
+    {
+      PatternRewriter::InsertionGuard g(rewriter);
```

Comment:
ubernit: IREE normally uses OpBuilder
```suggestion
      OpBuilder::InsertionGuard g(rewriter);
```